src/mame/drivers/spg2xx.cpp

```diff
@@ -1709,7 +1784,42 @@ void spg2xx_game_hotwheels_state::hotwheels(machine_config &config)
 	m_screen->set_refresh_hz(50);
 }
 
+uint16_t spg2xx_game_doraphone_state::porta_r(offs_t offset, uint16_t mem_mask)
+{
+	uint16_t data = m_io_p1->read() & 0xfff0;
+
+	switch (m_portb_data & 0xfe)
+	{
+	case 0xfe: data |= m_io_p1_fe->read() & 0x000f; break;
+	case 0xfc: data |= m_io_p1_fc->read() & 0x000f; break;
+	case 0xfa: data |= m_io_p1_fa->read() & 0x000f; break;
+	case 0xf6: data |= m_io_p1_f6->read() & 0x000f; break;
+	case 0xf4: data |= m_io_p1_f4->read() & 0x000f; break;
+	case 0xee: data |= m_io_p1_ee->read() & 0x000f; break;
+	case 0xde: data |= m_io_p1_de->read() & 0x000f; break;
+	case 0xbe: data |= m_io_p1_be->read() & 0x000f; break;
+	}
```

Comment:
This is clearly a key matrix.  Please test the column bits and then logical and with the appropriate rows rather than looking for specific values like this.

Reply:
sorry, not sure I understand what you mean here, how would you write this?

Reply:
See e.g. aed9f93f39fc107a8f54c2208bbea22e54954a2e

It looks like bits 1-6 are the active low row drive.  Reading with no rows driven (what you’re calling `P1_FE`) is a check for stuck bits, and `P1_F4` is the console reading the keys that are pressed on two rows at once.

Do something like:
```
...
    m_io_p1_rows(*this, "P1_ROW%u", 1U),
...
    required_ioport_array<6> m_io_p1_rows;
...
uint16_t spg2xx_game_doraphone_state::porta_r(offs_t offset, uint16_t mem_mask)
{
    uint16_t matrix = 0x000f;
    for (int b = 1; 6 >= b; ++b)
    {
        if (!BIT(offset, b))
            matrix &= m_io_p1_rows[b - 1]->read();
    }

    return matrix | (m_io_p1->read() & 0xfff0);
}
```

Then rename the real I/O ports to `P1_ROW1`..`P1_ROW6` and remove the non-existent `P1_FE` and `P1_F4`.

Reply:
Random example from a mahjong game: https://git.redump.net/mame/tree/src/mame/drivers/jongkyo.cpp#n179

Reply:
ah, interesting, that means the hidden test mode access must be different then, but I'll go with that